docs/content/stable/api/ysql/the-sql-language/statements/ddl_alter_table.md

```diff
@@ -222,68 +236,68 @@ It quietly succeeds. Now `\d children` shows that the foreign key constraint `ch
+
+The reason for the table rewrite is the clustered storage by primary key that YugabyteDB uses to store rows and indexes. Tables without a `PRIMARY KEY` have a hidden one underneath and rows are stored clustered on it. These rows need to be rewritten to use the newly added primary key column. 
 
-{{< warning >}}
-Adding a `PRIMARY KEY` constraint results in a full table rewrite and full rewrite of all indexes associated with the table.
-This happens because of the clustered storage by primary key that YugabyteDB uses to store rows and indexes.
-Tables without a `PRIMARY KEY` have a hidden one underneath and rows are stored clustered on it. The secondary indexes of the table
-link to this hidden `PRIMARY KEY`.
-While the tables and indexes are being rewritten, you may lose any modifications made to the table.
-For reference, the same semantics as [Alter type with table rewrite](#alter-type-with-table-rewrite) apply.
-{{< /warning >}}
 
 #### ALTER [ COLUMN ] *column_name* [ SET DATA ] TYPE *data_type* [ COLLATE *collation* ] [ USING *expression* ]
 
 Change the type of an existing column. The following semantics apply:
-- If data on disk is required to change, a full table rewrite is needed.
 - If the optional `COLLATE` clause is not specified, the default collation for the new column type will be used.
 - If the optional `USING` clause is not provided, the default conversion for the new column value will be the same as an assignment cast from the old type to the new type.
 - A `USING` clause must be included when there is no implicit assignment cast available from the old type to the new type.
 - Alter type is not supported for partitioned tables. See [#16980](https://github.com/yugabyte/yugabyte-db/issues/16980).
 - Alter type is not supported for tables with rules (limitation inherited from PostgreSQL).
 - Alter type is not supported for tables with CDC streams, or xCluster replication when it requires data on disk to change. See [#16625](https://github.com/yugabyte/yugabyte-db/issues/16625).
 
-##### Alter type without table-rewrite
+##### Table rewrites
 
-If the change doesn't require data on disk to change, concurrent DMLs to the table can be safely performed as shown in the following example:
+Altering a column’s type requires a [full table rewrite](#alter-table-operations-that-involve-a-table-rewrite) of the table and any indexes that contain this column when the underlying storage format changes or if the data changes.
 
+The following type changes are common cases where we require a table rewrite:
 
-```sql
-CREATE TABLE test (id BIGSERIAL PRIMARY KEY, a VARCHAR(50));
-ALTER TABLE test ALTER COLUMN a TYPE VARCHAR(51);
-```
+|     From         |  To               | Reason for table rewrite                                       |
```

Comment:
Maybe the Primary key page I linked above should have a link to this. How useful is this info? Is it buried here and should rather/additionally go in Architecture or develop or somewhere?